### ISSUE TITLE
Option type appears broken when interacting with cancelled .NET tasks

### ISSUE THREAD

**Issue by heron1** [dotnet/fsharp#16200]
Please provide a succinct description of the issue.

Setting a breakpoint at "0", we can see "b" incorrectly returns null instead of None from the following code:

```
open System.Net.Http
open System.Threading.Tasks

[<EntryPoint>]
let main str =
    let httpClient = new HttpClient()
    let b =
        task {
            try
               let! resp = httpClient.GetAsync($"invalid.invalid").ContinueWith(fun (t:Task<HttpResponseMessage>) -> t.Result)
               return Some 1
            with
                ex -> return None
        }
        |> Async.AwaitTask |> Async.RunSynchronously
    0
```

If however we create our own custom Option type, it works correctly:
```
open System.Net.Http
open System.Threading.Tasks

type OptionSelf<'T> =
    | Some of 'T
    | None

[<EntryPoint>]
let main str =
    let httpClient = new HttpClient()
    let b:OptionSelf<int> =
        task {
            try
               let! resp = httpClient.GetAsync($"invalid.invalid").ContinueWith(fun (t:Task<HttpResponseMessage>) -> t.Result)
               return Some 1
            with
                ex -> return None
        }
        |> Async.AwaitTask |> Async.RunSynchronously
    0
```

I notice 8 years ago this issue was reported as still existing with async instead of task based paralleism:
https://stackoverflow.com/questions/32806215/how-to-catch-exception-thrown-by-task-run-with-async-awaittask

I find it hard to believe until now this still hasn't been fixed. So is there something I'm missing, or is this really a bug until now that makes option types essentially unusable when interacting with .NET tasks?

**Expected behavior**

Option types should not be nullable when interacting with .NET tasks

**Actual behavior**

Option types can become null when interacting with .NET tasks

Provide any related information (optional):

* Operating system
.NET 7, F# 7, Windows 10

**Comment by En3Tho** [comment]
`None` is `null` in runtime. It's a special optimization and I don't see how it prevents you from doing the thing you want. It is also doesn't have anything to do with Tasks.

**Comment by heron1** [comment]
@En3Tho I see, I'm new to the language and did not know that. That was very confusing as a newcomer as I was relying on the debugger to show the correct types in an attempt to debug an issue I'm having, but fair enough.

For the Async.AwaitTask, my understanding is that this is necessary to convert the C# task to an F# async one, so that we can then do the synchronous Async.RunSynchronously?

**Comment by En3Tho** [comment]
@heron1 I see. Perhaps I came sort of aggressive. I'm sorry for that.

You need to use Async.AwaitTask in Async computation expressions (`async { ... }). With modern `task { ... }` builders you can do proper await just like you would do with an async: `task { let! x = someTask ... }`

Canonical way to wait for Tasks to finish is to call `.GetAwaiter().GetResult()`. This is the same as in C# if you're familiar with it

As for the DebugDisplay of None perhaps this can be improved. Like literally display "None" in debug mode to get things less confusing.

**Comment by heron1** [comment]
@En3Tho Not at all, my initial post perhaps contained some hints of frustration in it, as I spent a couple days attempting to resolve the issue by converting null to None, but your reply has helped clarify things for me, so thank you.

That's very useful to know about .GetAwaiter().GetResult() to keep things in the task based domain, I didn't come across that anywhere until now. And yes I agree if the debugger could show None instead of null I think it could alleviate a lot of confusion for newer developers to F#. I was looking in the wrong direction since I had a .IsNone code block not executing when the option was null, thinking that was the issue. Turns out it was something else.

As for this issue I believe it can be closed as .IsNone does indeed trigger correctly with null.
